diff --git a/nddiss2e.dtx b/nddiss2e.dtx
index 4f7b9df..4949315 100644
--- a/nddiss2e.dtx
+++ b/nddiss2e.dtx
@@ -13,7 +13,7 @@
 % \fi
 %
 %
-%  \CheckSum{1487}
+%  \CheckSum{1548}
 %  \CharacterTable
 %   {Upper-case    \A\B\C\D\E\F\G\H\I\J\K\L\M\N\O\P\Q\R\S\T\U\V\W\X\Y\Z
 %    Lower-case    \a\b\c\d\e\f\g\h\i\j\k\l\m\n\o\p\q\r\s\t\u\v\w\x\y\z
@@ -331,6 +331,15 @@
 % layout.
 % Do NOT use this option when preparing to submit it to the Graduate School.
 %
+% \DescribeMacro{showframe}
+% The |showframe| option shows visible frames for the text area and page.
+% This can be useful to ensure that all content is within the content area and
+% not running over into the margins.
+%
+% \DescribeMacro{linenumbers}
+% The |linenumbers| option enables the display of line numbers throughout the
+% document.
+%
 % \DescribeMacro{nocenter}
 % The |nocenter| option allows non-centered chapter titles.
 % Do NOT turn in your document this way to the Graduate School!
@@ -908,6 +917,7 @@
 \newif\ifnum@refs               \num@refstrue
 \newif\ifnatbib@refs            \natbib@refstrue
 \newif\ifcentered@chaptitle     \centered@chaptitletrue
+\newif\ifline@numbers           \line@numbersfalse
 \newif\if@ltfirstcaption
 %
 %    \end{macrocode}
@@ -1048,6 +1058,32 @@
   \PassOptionsToClass{twoside}{book} }%
 %
 %    \end{macrocode}
+% The |showframe| option uses the {\sffamily geometry} package to draw visible
+% frames for the text area and page. This is useful for checking that none of
+% document content runs outside the content area and into the margins.
+%    \begin{macrocode}
+\DeclareOption{showframe}{%
+  \ifdiss@review%
+    \PassOptionsToPackage{showframe}{geometry}%
+  \else%
+    \OptionNotUsed
+    \ClassWarningNoLine{nddiss2e}%
+      {Frame only shown in review mode; not showing frame}
+  \fi
+}
+%    \end{macrocode}
+% The |linenumbers| option enables line numbering throughout the document.
+%    \begin{macrocode}
+\DeclareOption{linenumbers}{%
+  \ifdiss@final%
+    \OptionNotUsed%
+    \ClassWarningNoLine{nddiss2e}%
+      {Line numbers are disabled in final mode.}%
+  \else%
+    \typeout{LINE NUMBERS}\line@numberstrue%
+  \fi
+}
+%    \end{macrocode}
 % All options other than those defined above are ignored and
 % a warning is printed
 % on the screen during compile-time. After processing all the options, the
@@ -1078,6 +1114,7 @@
   bookmarkstype=toc,%
   colorlinks=false,%
   pdfpagemode=UseOutlines]{hyperref}
+\RequirePackage[pass]{geometry}
 \RequirePackage{longtable}
 \RequirePackage[flushleft]{threeparttable}
 \RequirePackage[flushleft]{threeparttablex}
@@ -1174,6 +1211,39 @@
 \RequirePackage{metalogo}
 %
 %    \end{macrocode}
+% Here, if |linenumbers| is enabled, we load the \textsf{lineno} package and
+% start numbering with |\linenumbers| when (pdf)\LaTeX\space processes
+% |\begin{document}|. Additionally, we
+% \href{https://phaseportrait.blogspot.com/2007/08/lineno-and-amsmath-compatibility.html}{patch up}
+% a known incompatibility between the \textsf{lineno} and \textsf{amsmath} packages.
+%    \begin{macrocode}
+\ifline@numbers
+  \RequirePackage{lineno}
+  \AtBeginDocument{\linenumbers}
+  \newcommand*\@patch@lineno@amsmath[1]{%
+  \expandafter\let\csname old#1\expandafter\endcsname\csname #1\endcsname
+  \expandafter\let\csname oldend#1\expandafter\endcsname\csname end#1\endcsname
+  \renewenvironment{#1}%
+     {\linenomath\csname old#1\endcsname}%
+     {\csname oldend#1\endcsname\endlinenomath}%
+  }%
+  \AtBeginDocument{%
+    \@patch@lineno@amsmath{equation}%
+    \@patch@lineno@amsmath{equation*}%
+    \@patch@lineno@amsmath{align}%
+    \@patch@lineno@amsmath{align*}%
+    \@patch@lineno@amsmath{aligned}%
+    \@patch@lineno@amsmath{flalign}%
+    \@patch@lineno@amsmath{flalign*}%
+    \@patch@lineno@amsmath{alignat}%
+    \@patch@lineno@amsmath{alignat*}%
+    \@patch@lineno@amsmath{gather}%
+    \@patch@lineno@amsmath{gather*}%
+    \@patch@lineno@amsmath{multline}%
+    \@patch@lineno@amsmath{multline*}%
+  }
+\fi
+%    \end{macrocode}
 % Set the |\pagestyle| for the document to |plain| here
 % and define default spacing.
 %    \begin{macrocode}
